metronome/core/src/io/iohk/metronome/core/messages/RPCTracker.scala

```diff
@@ -0,0 +1,89 @@
+package io.iohk.metronome.core.messages
+
+import cats.implicits._
+import cats.effect.{Concurrent, Timer, Sync}
+import cats.effect.concurrent.{Ref, Deferred}
+import java.util.UUID
+import scala.concurrent.duration.FiniteDuration
+import scala.reflect.ClassTag
+
+/** `RPCTracker` can be used to register outgoing requests and later
+  * match them up with incoming responses, thus turning the two independent
+  * messages into a `Kleisli[F, Request, Option[Response]]`, where a `None`
+  * result means the request timed out.
+  */
+class RPCTracker[F[_]: Timer: Concurrent, M](
+    deferredMapRef: Ref[F, Map[UUID, RPCTracker.Entry[F, _]]],
+    defaultTimeout: FiniteDuration
+) {
+  import RPCTracker.Entry
+
+  def register[
+      Req <: RPCMessageCompanion#Request,
+      Res <: RPCMessageCompanion#Response
+  ](
+      request: Req,
+      timeout: FiniteDuration = defaultTimeout
+  )(implicit
+      ev1: Req <:< M,
+      ev2: RPCPair.Aux[Req, Res],
```

Comment:
I'm a bit confused:
- how is this evidence used?
- why do we need `Aux`? Wouldn't `type RPCPair[Req, Res]` suffice?
- is there going to be one tracker for all sync messages, or one for each concrete pair like `GetBlock*`?

Reply:
> how is this evidence used?

It's here so the compiler can check that if we send a given `Req` type then we are right to expect a `Res` to be returned. Not just that, but should also allow the compiler to infer the return type based on just the `request` parameter, without explicitly having to define both generic types for `register`. 

> why do we need Aux? Wouldn't type RPCPair[Req, Res] suffice?

Every request has one possible response (although I suppose the same response could be paired with multiple responses), so `RPCPair` is just defined in terms of `Req`, with `Res` being a dependant type. Having two parameters would allow the same request to have multiple responses.

The `Aux` is needed so we can match against the two generic types of the `register` method. 

Otherwise it could be like this:
```scala
def register[
      Req <: RPCMessageCompanion#Request
  ](
      request: Req,
      timeout: FiniteDuration = defaultTimeout
  )(implicit
      ev1: Req <:< M,
      ev2: RPCPair[Req]
  ): F[F[Option[ev2.Res]]]
```
But then I couldn't have a `ct: ClassTag[Res]` because it cannot be matched against a dependent type of a parameter, `ev2`.

> is there going to be one tracker for all sync messages, or one for each concrete pair like GetBlock*?

Just one tracker per "service", like the `SyncService`.

Reply:
Added comments.

Reply:
Thanks for the explanation